[SPARK-6440][CORE]Handle IPv6 addresses properly when constructing URI

nyaapa opened the pull request.

nyaapa committed:
[SPARK-6440][CORE]Handle IPv6 addresses properly when constructing URI

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -806,15 +808,20 @@ private[spark] object Utils extends Logging {
+          if ( !addresses.isEmpty ) {
+            val addr = addresses.find( _.isInstanceOf[Inet4Address] ).getOrElse(addresses.head)
+            // because of Inet6Address.toHostName may add interface at the end if it knows about it
+            val strippedAddress = InetAddress.getByAddress(addr.getAddress)
```
Tell me more about what this fixes?

nyaapa replied:
It's all about finding our external IP address: old logic was to use external ipv4 || 127.0.0.1. 
That transformes it to ipv4 || ipv6 || 127.0.0.1

Whenewer we don't have any ipv4 it's better to use ipv6 than 127.0.0.1

srowen replied:
OK, though I don't think that was the thrust of this JIRA, which was simply to format IPv6 addresses correctly. I think the reasoning makes some sense though.

Nits: can you `filterNot(addr => addr.isLinkLocalAddress || addr.isLoopbackAddress)`
and `if (addresses.notEmpty) {`? and `find(_.isInstanceOf[Inet4Address])`? Just questions of style in this block.

It might be worth a comment to the effect of your reply above here.

nyaapa replied:
Ok, i agree with that. Should i make additional commit or change existing one?
*```nonEmpty```

srowen replied:
Just push more commits to this branch

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -845,6 +852,13 @@ private[spark] object Utils extends Logging {
+  /**
+   * Get the local machine's URI.
+   */
+  def localHostURI(): String = {
```
Is this intended to be an IP address or host name? looks like IP address. But then some of the changes above are changing a host name to an IP address. That might be OK but it's not obvious? what's the motivation there.

nyaapa replied:
It is named ```localHostURI``` in compliance with ```localHostName``` function name.
The main idea is to use ip if and only if we don't have ```customHostname``` ( like in ```localHostName``` ).

srowen replied:
Got it. I'm wondering when you _wouldn't_ want to use the "URI safe" name? they're identical in almost every case. Hm. I suppose the safer thing is to only format for URIs where, well, creating URIs. I'll have to dig into this more but generally want to minimize the change here and narrowly address the formatting issue

nyaapa replied:
I don't understand how to use "URI safe" name, something like ```localHostNameURISafe``` or what?
It's good to format URIs only in places where we need it, but currently we have more than one bug because of it.
However we can use something like  ```InetAddresses.toUriString(InetAddress.getByName(Utils.localHostName()))``` to get URI in place without fixing Utils module, but i think that special URI method may prevent URI-based bugs in future due to its existence.

srowen commented:
ok to test

SparkQA commented:
[Test build #29936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29936/consoleFull) for   PR 5424 at commit [`84763d7`](https://github.com/apache/spark/commit/84763d7b4ec94d1b792b93a0de52bcd20f3e7eee).

SparkQA commented:
[Test build #29936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29936/consoleFull) for   PR 5424 at commit [`84763d7`](https://github.com/apache/spark/commit/84763d7b4ec94d1b792b93a0de52bcd20f3e7eee).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29936/
Test FAILed.

nyaapa committed:
[SPARK-6440][CORE] style fixes and use getHostAddress instead of getHostName

nyaapa commented:
Use ```getHostAddress``` instead of ```getHostName```, because of ```getHostName``` may return additional information like ```95.108.174.236-red.dhcp.yndx.net```.

SparkQA commented:
[Test build #29940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29940/consoleFull) for   PR 5424 at commit [`2098081`](https://github.com/apache/spark/commit/209808105a8a0ef5749512dbc17cd8cde63b9bf0).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29940/
Test FAILed.

shaneknapp commented:
jenkins, test this please

SparkQA commented:
[Test build #29943 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29943/consoleFull) for   PR 5424 at commit [`2098081`](https://github.com/apache/spark/commit/209808105a8a0ef5749512dbc17cd8cde63b9bf0).

SparkQA commented:
[Test build #29943 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29943/consoleFull) for   PR 5424 at commit [`2098081`](https://github.com/apache/spark/commit/209808105a8a0ef5749512dbc17cd8cde63b9bf0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29943/
Test FAILed.

nyaapa commented:
Seems like jenkins was killed again.

shaneknapp commented:
whoops, i missed retriggering this build.  :)

jenkins, test this please

srowen commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #29962 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29962/consoleFull) for   PR 5424 at commit [`2098081`](https://github.com/apache/spark/commit/209808105a8a0ef5749512dbc17cd8cde63b9bf0).

SparkQA commented:
[Test build #29962 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29962/consoleFull) for   PR 5424 at commit [`2098081`](https://github.com/apache/spark/commit/209808105a8a0ef5749512dbc17cd8cde63b9bf0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29962/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #29974 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29974/consoleFull) for   PR 5424 at commit [`2098081`](https://github.com/apache/spark/commit/209808105a8a0ef5749512dbc17cd8cde63b9bf0).

SparkQA commented:
[Test build #29974 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29974/consoleFull) for   PR 5424 at commit [`2098081`](https://github.com/apache/spark/commit/209808105a8a0ef5749512dbc17cd8cde63b9bf0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29974/
Test PASSed.

srowen commented:
I think I've looked at 10 different versions of 2 PRs for this JIRA now, and all seem to be changing or adding something slightly different than what's essential here. Let me describe what I think needs to be done and why and we can settle that, then implement.

The code needs a local IP address in many cases. That's localIpAddress. (Separately, your update to find an IPv6 address sounds OK.)

localIpAddressHostname and the getAddressHostName() method actually look pointless. They make a String into an InetAddress and then ask for the exact same text representation back as a String. Personally I'd make localIpAddress an InetAddress, as you've done. I'd remove localIpAddressHostname and replace with localHostName() where a String is needed.

... because then customHostname is consistently applied, right? I think that must be the intent?

But then, localIpAddress itself doesn't have much value. It is always used for its string rep anyway. So, localHostName() as well? it's the only thing that needs to be non-private.

Back to the original issue: where localHostName() would then be used in a URI, it can be run through the utility method in InetAddresses. I don't think that we should instead have more methods in Utils just for adding a call to that method.

Finally, there are instances in SparkSQLEnv, KinesisReceiver and TestUtils that use InetAddress.getLocalHost directly. These should also use localHostName().

Maybe there's a reason that doesn't 100% work but it seems a lot less messy than the 3-5 ways code is reasoning about the local IP address string now.

What say, or, should I try it in my own PR?

nyaapa commented:
Ok, i'll try to:
* replace ```localIpAddressHostname``` in ```Utils.localHostName with``` ```localIpAddress.getHostAddress```
* replace ```localIpAddressHostname``` usage with ```localHostName```
* remove ```localIpAddressHostname``` and ```getAddressHostName```
* make ```localIpAddress``` private
* fix SparkSQLEnv, KinesisReceiver and TestUtils 

today/tomorrow.

>> Back to the original issue: where localHostName() would then be used in a URI, it can be run through the utility method in InetAddresses.
>> I don't think that we should instead have more methods in Utils just for adding a call to that method.

I didn't find any method in InetAddresses like ```escapeHostname(String) => String```. ```InetAddresses.toUriString(InetAddress.getByName(Utils.localHostName()))```, but this looks pointless like ```localIpAddressHostname = InetAddress.getByName(localIpAddress).getHostName```. 
And there we have an option: 
1. make ```localIpAddress``` private, remove ```localIpAddressURI``` variable and add ```def localHostURI() = customHostname.getOrElse(InetAddresses.toUriString(localIpAddress))```
2. ask everyone to use ```InetAddresses.toUriString(InetAddress.getByName(Utils.localHostName()))``` 

imho additional method with private variable is better. Also i don't like conversion from ```String``` to ```InetAddress```

srowen commented:
`InetAddresses.toUriString` accepts a `String` and returns an `InetAddress`, so it does not need its input changed. Yes you have to call `getHostName` on the output.

nyaapa commented:
```InetAddresses.toUriString``` is ```(InetAddress) => String```, maybe you meant ```forUriString(String) => InetAddress``` or ```forString(String) => InetAddress``` ?
for ```forUriString```
```
scala> val ipv6 = InetAddress.getByAddress(NetworkInterface.getNetworkInterfaces.toList.apply(0).getInetAddresses.toList.apply(0).getAddress).getHostAddress
ipv6: String = fe80:0:0:0:3623:87ff:fe6a:423c

scala> InetAddresses.forUriString(ipv6)
java.lang.IllegalArgumentException: Not a valid URI IP literal: 'fe80:0:0:0:3623:87ff:fe6a:423c'
  at com.google.common.net.InetAddresses.forUriString(InetAddresses.java:501)
  ... 33 elided
```
And for ```forString``` i still need ```InetAddresses.toUriString```
```
scala> val ipv6 = InetAddress.getByAddress(NetworkInterface.getNetworkInterfaces.toList.apply(0).getInetAddresses.toList.apply(0).getAddress).getHostAddress
ipv6: String = fe80:0:0:0:3623:87ff:fe6a:423c

scala> InetAddresses.forString(ipv6).getHostName
res35: String = fe80:0:0:0:3623:87ff:fe6a:423c

scala> InetAddresses.forString(ipv6).getHostAddress
res36: String = fe80:0:0:0:3623:87ff:fe6a:423c

scala> InetAddresses.toUriString(InetAddresses.forString(ipv6))
res37: String = [fe80::3623:87ff:fe6a:423c]
```

srowen commented:
Oh gosh yeah looking at this the wrong way around of course. Aren't there just two places where the URI logic is needed? I think calling `InetAddresses.toUriString(InetAddress.getByName(Utils.localHostName()))` in two places isn't so bad. A new method is OK too if you rprefer; I suppose ... `localHostNameForURI()` or something to make the relation clear? Then we have just two methods for accessing 'the local host name' with a clear role.

nyaapa commented:
Ok, ```localHostNameForURI``` sounds reasonable.
About ```network/common/src/test/java/org/apache/spark/network/TestUtils.java```: there is just only one method -- ```getLocalHost() => String```, i can delete the whole class and use ```Utils.localHostname``` instead, but maybe using scala from java it's not a good practice?

srowen commented:
I'd just leave `TestUtils` for purposes here, but can make it call through to `Utils`.

nyaapa committed:
[SPARK-6440][CORE] Remove Utils.localIpAddressHostname, Utils.localIpAddressURI and Utils.getAddressHostName;
make Utils.localIpAddress private;
rename Utils.localHostURI into Utils.localHostNameForURI;
use Utils.localHostName in org.apache.spark.streaming.kinesis.KinesisReceiver and org.apache.spark.sql.hive.thriftserver.SparkSQLEnv

SparkQA commented:
[Test build #30091 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30091/consoleFull) for   PR 5424 at commit [`6b717aa`](https://github.com/apache/spark/commit/6b717aa07f29087c25cdc75bd3c46f273d10a94b).

SparkQA commented:
[Test build #30091 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30091/consoleFull) for   PR 5424 at commit [`6b717aa`](https://github.com/apache/spark/commit/6b717aa07f29087c25cdc75bd3c46f273d10a94b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30091/
Test PASSed.

srowen commented:
LGTM. This simplifies and standardizes the local host name logic across the code, fixes the original problem of IPv6 addresses in URIs, chooses a better local address in some IPv6 cases. Of course, making things consistent means slight behavior changes in some cases -- for example, `customHostName` now applies in cases it didn't before, but this actually seems more correct. I've eyeballed every change and it looks logical, and the tests pass.
